File: src/main/java/com/freya02/botcommands/api/localization/LocalizationMap.java
package com.freya02.botcommands.api.localization;

import com.freya02.botcommands.api.localization.providers.LocalizationMapProvider;
import org.jetbrains.annotations.NotNull;

import java.util.Locale;
import java.util.Map;

/**
 * Interface specifying getters for the effective locale and template map
 */
public interface LocalizationMap {
	/**
	 * Returns the effective locale.
	 * <br>This locale must be the one of the file which has been loaded successfully.
	 * <br><b>It does not have to be the locale passed to {@link LocalizationMapProvider#getBundle(String, Locale)}</b>.
	 *
	 * @return The effective locale for this localization map
	 */
	@NotNull Locale effectiveLocale();

	/**
	 * Returns the localization template map.
	 * <br>The key is the localization path, such as {@code my_command.name},
	 * and the value is a LocalizationTemplate of your choice, possibly {@link DefaultLocalizationTemplate}.
	 *
	 * @return The localization template map
	 */
	@NotNull Map<String, ? extends LocalizationTemplate> templateMap();
}

File: src/main/kotlin/com/freya02/botcommands/api/localization/LocalizationMap.kt
package com.freya02.botcommands.api.localization

import com.freya02.botcommands.api.localization.providers.LocalizationMapProvider
import org.jetbrains.annotations.UnmodifiableView
import java.util.*

/**
 * Common interface to retrieve [localization templates][LocalizationTemplate] from a path.
 */
interface LocalizationMap {
    /**
     * Returns the effective locale of this localization map.
     *
     * **Note:** this doesn't need to be the locale passed by [LocalizationMapProvider.fromBundleOrParent].
     */
    val effectiveLocale: Locale

    /**
     * Returns an unmodifiable set of keys this localization map contains,
     * or `null` if unsupported.
     */
    val keys: @UnmodifiableView Set<String>?

    /**
     * Returns the [LocalizationTemplate] with the corresponding path,
     * or `null` if there is no such entry.
     */
    operator fun get(path: String): LocalizationTemplate?
}

fun createDelegated(current: LocalizationMap?, parent: LocalizationMap): LocalizationMap {
    if (current == null) return parent

    return object : LocalizationMap {
        override val effectiveLocale: Locale
            get() = current.effectiveLocale

        override val keys: Set<String>?
            get() = current.keys

        override fun get(path: String): LocalizationTemplate? =
            current[path] ?: parent[path]
    }
}